### PR TITLE
Constified `array::from_fn` and `ptr::drop_in_place`

### PR DESCRIPTION
**Pull request by fee1-dead**
- [x] Add unit test
 - [x] Add tracking issue to feature

### CODE REVIEW

File: library/core/src/ops/try_trait.rs [hunk 1]
```diff
@@ -384,8 +385,17 @@ impl<T> NeverShortCircuit<T> {
     /// This is useful for implementing infallible functions in terms of the `try_` ones,
     /// without accidentally capturing extra generic parameters in a closure.
     #[inline]
-    pub fn wrap_mut_1<A>(mut f: impl FnMut(A) -> T) -> impl FnMut(A) -> NeverShortCircuit<T> {
-        move |a| NeverShortCircuit(f(a))
+    pub const fn wrap_mut_1<A>(
+        mut f: impl ~const FnMut(A) -> T + ~const Destruct,
+    ) -> impl ~const FnMut(A) -> NeverShortCircuit<T> + ~const Destruct {
```
Comment (scottmcm) [inline library/core/src/ops/try_trait.rs:390]: Nice to see that `impl ~const` works now.

Comment (est31) [reply library/core/src/ops/try_trait.rs:390]: I think this was #105725

File: tests/ui/consts/issue-102117.rs [hunk 2]
```diff
@@ -17,10 +17,9 @@ impl VTable {
             &VTable {
                 layout: Layout::new::<T>(),
                 type_id: TypeId::of::<T>(),
-                //~^ ERROR the parameter type `T` may not live long enough
-                //~| ERROR the parameter type `T` may not live long enough
                 drop_in_place: unsafe {
                     transmute::<unsafe fn(*mut T), unsafe fn(*mut ())>(drop_in_place::<T>)
+                    //~^ ERROR can't drop `T` in const contexts
```
Comment (cuviper) [inline tests/ui/consts/issue-102117.rs:22]: Isn't this a bad error? We're not actually *calling* the function that needs const drop, and this breaks stable code:

```rust
pub const fn get_drop<T>() -> unsafe fn(*mut T) {
    core::ptr::drop_in_place::<T>
}
```

Comment (cuviper) [reply tests/ui/consts/issue-102117.rs:22]: Similarly for `array::from_fn`:

```rust
pub const fn array_from_fn() -> fn(fn(usize) -> ()) -> [(); 42] {
    core::array::from_fn
}
```

... works on stable, but fails here for lack of `~const FnMut` on the function pointer.

Comment (cuviper) [reply tests/ui/consts/issue-102117.rs:22]: Does this go back to the previous error after #109557?

Comment (fee1-dead) [reply tests/ui/consts/issue-102117.rs:22]: The const-eval working group has been discussing the status of const traits in the compiler as well as the standard library. We came to a consensus that all uses of `~const` in the standard library should be temporary removed to allow for a clean impl of it to land in the compiler. See [this zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/.60const.20Trait.60.20removal.20or.20rework).

**Pull request closed without merging**